src/snowflake/cli/api/project/definition_conversion.py

```diff
@@ -380,33 +450,118 @@ def _convert_templates_in_files(
             definition_v2.get_entities_by_type(
                 ApplicationPackageEntityModel.get_type()
             ).values()
         )[0]
-        converted_post_deploy_hooks = _convert_package_script_files(
-            project_root, pkg.scripts, pkg_entity, in_memory
-        )
-        if pkg_entity.meta is None:
-            pkg_entity.meta = MetaField()
-        if pkg_entity.meta.post_deploy is None:
-            pkg_entity.meta.post_deploy = []
-        pkg_entity.meta.post_deploy += converted_post_deploy_hooks
+
+        # Convert templates in artifacts by passing them through the TemplatesProcessor
+        # but providing a context that maps v1 template references to the equivalent v2
+        # references instead of resolving to literals
+        # For example, replacement_template_context might look like
+        # {
+        #     "ctx": {
+        #         "native_app": {
+        #             "bundle_root": "<% ctx.entities.pkg.bundle_root %>",
+        #             "deploy_root": "<% ctx.entities.pkg.deploy_root %>",
+        #             "application": {
+        #                 "name": "<% ctx.entities.app.identifier %>",
+        #             }
+        #             and so on...
+        #          }
+        #     }
+        # }
+        # We only convert files on-disk if the "templates" processor is used in the artifacts
+        # and if we're doing a permanent conversion. If we're doing an in-memory conversion,
+        # the CLI global template context is already populated with the v1 definition, so
+        # we don't want to convert the v1 template references in artifact files
+        if not in_memory and any(
+            processor.name == TEMPLATES_PROCESSOR
+            for artifact in pkg_model.artifacts
+            for processor in artifact.processors
+        ):
+            metrics.set_counter(CLICounterField.TEMPLATES_PROCESSOR, 1)
+
+            # Create a temporary directory to hold the expanded templates,
+            # as if a bundle step had been run but without affecting any
+            # files on disk outside of the artifacts we want to convert
+            with tempfile.TemporaryDirectory() as d:
+                deploy_root = Path(d)
+                bundle_ctx = BundleContext(
+                    package_name=pkg_model.identifier,
+                    artifacts=pkg_model.artifacts,
+                    project_root=project_root,
+                    bundle_root=project_root / pkg_model.bundle_root,
+                    deploy_root=deploy_root,
+                    generated_root=(
+                        project_root / deploy_root / pkg_model.generated_root
+                    ),
+                )
+                template_processor = TemplatesProcessor(bundle_ctx)
+                for artifact in na.artifacts:
+                    for processor in artifact.processors:
+                        if processor.name == TEMPLATES_PROCESSOR:
```

Comment:
If we're inside the outer if block that checks if the templates processor is present, why do we need this loop and the check again?

Reply:
yeah that's correct, the first check bypasses everything if none of the artifacts use the templates processor, this check makes us apply the processor to only the artifacts that have opted-in to it (since other artifact entries may have files with tags that they don't want templated)

Reply:
small nit then is that maybe we can build the list of artifacts that opted-in to the processor, then in the check we can make sure that's not empty, and loop over it for the conversion

(e.g. something like `[artifact for artifact in pkg_model.artifacts for processor in artifact.processors if processor.name == TEMPLATES_PROCESSOR]`)

to reduce the nesting a bit and make it easier to reason about, wdyt?

Reply:
yeah i can do that